elyra/pipeline/pipeline_definition.py

```diff
@@ -0,0 +1,281 @@
+    def __init__(self, pipeline_definition: Dict, node: Dict):
+        super().__init__(pipeline_definition, node)
+
+    @property
+    def version(self) -> int:
+        return int(self._node['app_data'].get('version'))
+
+    @property
+    def runtime(self) -> str:
+        return self._node['app_data'].get('runtime') or 'local'
+
+    @property
+    def type(self):
+        type_description_to_type = {'Kubeflow Pipelines': 'kfp',
+                                    'Apache Airflow': 'airflow',
+                                    'Generic': 'generic'}
+
+        pipeline_type_description = self._node['app_data']['properties'].get('runtime')
+        if pipeline_type_description not in type_description_to_type.keys():
+            raise ValueError(f'Unsupported pipeline runtime: {pipeline_type_description}')
+        return type_description_to_type[pipeline_type_description]
+
+    @property
+    def nodes(self) -> list:
+        if 'nodes' not in self._node:
+            raise ValueError("Pipeline is missing 'nodes' field.")
+        elif len(self._node['nodes']) == 0:
+            raise ValueError("Pipeline has zero length 'nodes' field.")
+
+        nodes: list = list()
+        for node in self._node['nodes']:
+            nodes.append(Node(self._pipeline_definition, node))
+
+        return nodes
+
+    def get_property(self, key: str, default_value=None) -> Any:
+        self._validate()
+        return self._node['app_data']['properties'].get(key, default_value)
+
+    def set_property(self, key: str, value: Any):
+        if not key:
+            raise ValueError("Key is required")
+
+        if not value:
+            raise ValueError("Value is required")
+
+        self._node['app_data']['properties'][key] = value
+
+    def validate(self) -> list:
+        validation_issues = list()
+        if 'properties' not in self._node['app_data']:
+            validation_issues.append("Node is missing 'properties' field.")
+        elif len(self._node['app_data']['properties']) == 0:
+            validation_issues.append("Pipeline has zero length 'properties' field.")
+
+        return validation_issues
+
+
+class Node(NodeBase):
+    def __init__(self, pipeline_definition: Dict, node: Dict):
+        super().__init__(pipeline_definition, node)
+
+    @property
+    def type(self) -> str:
+        return self._node.get('type')
+
+    @property
+    def op(self) -> str:
+        return self._node.get('op')
+
+    @property
+    def label(self) -> str:
+        return self._node['app_data']['ui_data'].get('label', self._node['app_data']['label'])
+
+    def get_component_parameter(self, key: str, default_value=None) -> Any:
+        self._validate()
+        return self._node['app_data']['component_parameters'].get(key, default_value)
+
+    def set_component_parameter(self, key: str, value: Any):
+        if not key:
+            raise ValueError("Key is required")
+
+        if not value:
+            raise ValueError("Value is required")
```

Comment:
Third occurrence of this pattern.  Perhaps a `_vallidate_key_value()` helper in `NodeBase`?
```python
def _validate_key_value(self, key: str, value: Any) -> None:
        if not key:
            raise ValueError("Key is required")

        if not value:
            raise ValueError("Value is required")
```